packages/medusa/src/services/batch-job.ts

```diff
@@ -71,6 +144,235 @@ class BatchJobService extends TransactionBaseService<BatchJobService> {
+    // TODO validate context with batch job strategy corresponding to the given type
+
+    return context
+  }
+
+  async create(data: BatchJobCreateProps): Promise<BatchJob> {
+    return await this.atomicPhase_(async (manager) => {
+      const batchJobRepo: BatchJobRepository = manager.getCustomRepository(
+        this.batchJobRepository_
+      )
+
+      const validatedContext = await this.validateBatchContext(
+        data.type,
+        data.context
+      )
+
+      const toCreate = {
+        ...data,
+        context: validatedContext,
+      } as DeepPartial<BatchJob>
+
+      const batchJob = batchJobRepo.create(toCreate)
+      const result = await batchJobRepo.save(batchJob)
+
+      await this.eventBus_
+        .withTransaction(manager)
+        .emit(BatchJobService.Events.CREATED, {
+          id: result.id,
+        })
+
+      return result
+    })
+  }
+
+  async update(
+    batchJobId: string,
+    data: BatchJobUpdateProps
+  ): Promise<BatchJob> {
+    return await this.atomicPhase_(async (manager) => {
+      const batchJobRepo: BatchJobRepository = manager.getCustomRepository(
+        this.batchJobRepository_
+      )
+
+      let batchJob = await this.retrieve(batchJobId)
+
+      const { context, ...rest } = data
+      if (context) {
+        batchJob.context = await this.validateBatchContext(
+          batchJob.type,
+          context
+        )
+      }
+
+      Object.keys(rest)
+        .filter((key) => typeof rest[key] !== `undefined`)
+        .forEach((key) => {
+          batchJob[key] = rest[key]
+        })
+
+      batchJob = await batchJobRepo.save(batchJob)
+
+      await this.eventBus_
+        .withTransaction(manager)
+        .emit(BatchJobService.Events.UPDATED, {
+          id: batchJob.id,
+        })
+
+      return batchJob
+    })
+  }
+
+  protected async updateStatus(
+    batchJobOrId: BatchJob | string,
+    status: BatchJobStatus
+  ): Promise<BatchJob | never> {
+    const transactionManager = this.transactionManager_ ?? this.manager_
+    let batchJob: BatchJob = batchJobOrId as BatchJob
+    if (typeof batchJobOrId === "string") {
+      batchJob = await this.retrieve(batchJobOrId)
+    }
+
+    const { entityColumnName, eventType } =
+      this.batchJobStatusMapToProps.get(status) || {}
+
+    if (!entityColumnName || !eventType) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `Unable to update the batch job status from ${batchJob.status} to ${status}. The status doesn't exist`
+      )
+    }
+
+    batchJob[entityColumnName] = new Date()
+
+    const batchJobRepo = transactionManager.getCustomRepository(
+      this.batchJobRepository_
+    )
+    batchJob = await batchJobRepo.save(batchJob)
+    batchJob.loadStatus()
+
+    this.eventBus_.withTransaction(transactionManager).emit(eventType, {
+      id: batchJob.id,
+    })
+
+    return batchJob
+  }
+
+  async setConfirm(batchJobOrId: string | BatchJob): Promise<BatchJob | never> {
```

Comment:
todo: would keep this named `confirm` as that is a pretty clear action. The method names that I would suggest prefixing with `set` are only the ones that didn't already describe an action e.g. `ready` or `preProcessing`, etc. 

Will highlight changes to be made below